sequencer/src/block/full_payload/ns_table.rs

```diff
@@ -0,0 +1,229 @@
+//! Types related to a namespace table.
+//!
+//! All code that needs to know the binary format of a namespace table is
+//! restricted to this file.
+use crate::{
+    block::{
+        full_payload::payload::PayloadByteLen,
+        namespace_payload::NsPayloadRange,
+        uint_bytes::{
+            bytes_serde_impl, u64_from_bytes, u64_to_bytes, usize_from_bytes, usize_to_bytes,
+        },
+    },
+    NamespaceId,
+};
+use committable::{Commitment, Committable, RawCommitmentBuilder};
+use hotshot_types::traits::EncodeBytes;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::{collections::HashSet, sync::Arc};
+
+/// Byte lengths for the different items that could appear in a namespace table.
+const NUM_NSS_BYTE_LEN: usize = 4;
+const NS_OFFSET_BYTE_LEN: usize = 4;
+const NS_ID_BYTE_LEN: usize = 8;
+
+#[repr(transparent)]
+#[derive(Clone, Debug, Default, Deserialize, Eq, Hash, PartialEq, Serialize)]
+#[serde(transparent)]
+pub struct NsTable(#[serde(with = "base64_bytes")] Vec<u8>);
+
+impl NsTable {
+    /// Search the namespace table for the ns_index belonging to `ns_id`.
+    pub fn find_ns_id(&self, ns_id: &NamespaceId) -> Option<NsIndex> {
+        self.iter().find(|index| self.read_ns_id(index) == *ns_id)
+    }
+
+    /// Iterator over all unique namespaces in the namespace table.
+    pub fn iter(&self) -> impl Iterator<Item = NsIndex> + '_ {
+        NsIter::new(self)
+    }
+
+    /// Read the namespace id from the `index`th entry from the namespace table.
+    ///
+    /// `index` is not checked. Use [`Self::in_bounds`] as needed.
+    ///
+    /// TODO I want to restrict visibility to `pub(crate)` or lower but this
+    /// method is currently used in `nasty-client`.
+    pub fn read_ns_id(&self, index: &NsIndex) -> NamespaceId {
+        let start = index.0 * (NS_ID_BYTE_LEN + NS_OFFSET_BYTE_LEN) + NUM_NSS_BYTE_LEN;
+
+        // TODO hack to deserialize `NamespaceId` from `NS_ID_BYTE_LEN` bytes
+        NamespaceId::from(u64_from_bytes::<NS_ID_BYTE_LEN>(
+            &self.0[start..start + NS_ID_BYTE_LEN],
+        ))
+    }
+
+    /// Does the `index`th entry exist in the namespace table?
+    pub fn in_bounds(&self, index: &NsIndex) -> bool {
+        // The number of entries in the namespace table, including all duplicate
```

Comment:
Can't the replicas ensure the namespaces are not duplicated?

Reply:
They could, and perhaps they will. This design was originally conceived on the assumption that replicas would not enforce anything, and a namespace table could have arbitrary bytes. I'd like to maintain that property as long as it is reasonable to do so, just in case we need it again in the future.

Reply:
I see. Thank you for the clarification.